docs: Adding Best practices for Deploying and Running Langflow in production

nidhimishra-datastax opened the pull request:
- Added Added Best practices for Deploying Langflow in Production `docs/docs/Deployment/deployment-prod-best-practices.md`
- Added instructions for setting up Langflow Dev env `docs/docs/Deployment/deployment-kubernetes-dev.md`
- Added instructions for setting up Langflow Prod env `docs/docs/Deployment/deployment-kubernetes-prod.md`
- Updated the sidebar configuration to reflect the three new files `docs/sidebars.js`
- Diagram for Segregation of Dev and Prod environments `docs/static//img/langflow-env.png`
- Diagram for Langflow Reference Architecture `docs/static//img/langflow-reference-architecture.png`

nidhimishra-datastax commented:
@mendonk Please review.

msmygit commented on docs/docs/Deployment/@:
```diff
@@ -0,0 +1,621 @@
+---
```
@nidhimishra-datastax is the filename `@` here?

nidhimishra-datastax commented:
@mendonk I have updated the changes to be Dark mode friendly and renamed the Best practices doc to Langflow Architecture and Best Practices for clarity
<img width="1512" alt="Screenshot 2025-04-11 at 09 37 07" src="https://github.com/user-attachments/assets/64284bf6-805c-4328-bd65-87ed434e51c6" />
<img width="1497" alt="Screenshot 2025-04-11 at 09 37 17" src="https://github.com/user-attachments/assets/e4696916-4b09-4f28-a073-577a0b1bab51" />

mendonk commented:
Thank you @nidhimishra-datastax , myself and docs team will review and I will let you know when your changes are live.

mendonk commented on docs/docs/Deployment/deployment-prod-best-practices.md:
```diff
@@ -0,0 +1,56 @@
+
+## Environment isolation
+
+It is recommended to deploy and run two separate environments for Langflow.
```
```suggestion
It is recommended to deploy and run two separate environments for Langflow, with one environment reserved for Development use and another for Production use.
```

mendonk commented on docs/docs/Deployment/deployment-prod-best-practices.md:
```diff
@@ -0,0 +1,56 @@
+
+It is recommended to deploy and run two separate environments for Langflow.
+
+One environment to be reserved for Development use and another for Production use.
```
```suggestion
```

mendonk commented on docs/docs/Deployment/deployment-prod-best-practices.md:
```diff
@@ -0,0 +1,56 @@
+
+## Why is it important to have separate deployments?
+
+This separation is designed to enhance security, optimize resource allocation, and streamline management. Understanding the rationale behind these deployment options help you make informed decisions about how to best deploy and manage your applications.
```
```suggestion
This separation is designed to enhance security, optimize resource allocation, and streamline management.
```

mendonk commented on docs/docs/Deployment/deployment-prod-best-practices.md:
```diff
@@ -0,0 +1,56 @@
+  * **Access Control**: Different security policies and access controls can be applied to each environment. Developers may require broader access in the IDE for testing and debugging, while the runtime environment can be locked down with stricter security measures.
+  * **Reduced Attack Surface**: The runtime environment is configured to only include essential components, reducing the attack surface and potential vulnerabilities.
+* **Resource Allocation**
+  * **Optimized Resource Usage and cost efficiency**: By separating the two, we can allocate resources more effectively. Additionally, each flow can be deployed independently, providing fine-grained resource control.
```
```suggestion
  * **Optimized Resource Usage and cost efficiency**: By separating the two environments, you can allocate resources more effectively. Each flow can be deployed independently, providing fine-grained resource control.
```

mendonk reviewed (commented):
Suggesting a few changes for style and clarity.

KimberlyFields commented on docs/docs/Deployment/deployment-prod-best-practices.md:
```diff
@@ -0,0 +1,55 @@
+
+## Environment isolation
+
+It is recommended to deploy and run two separate environments for Langflow, with one environment reserved for Development use and another for Production use.
```
```suggestion
It is recommended to deploy and run two separate environments for Langflow, with one environment reserved for development use and another for production use.
```

KimberlyFields commented on docs/docs/Deployment/deployment-prod-best-practices.md:
```diff
@@ -0,0 +1,55 @@
+It is recommended to deploy and run two separate environments for Langflow, with one environment reserved for Development use and another for Production use.
+
+
+![Langflow Environments](/img/langflow-env.png)
```
```suggestion
![Langflow environments](/img/langflow-env.png)
```

KimberlyFields commented on docs/docs/Deployment/deployment-prod-best-practices.md:
```diff
@@ -0,0 +1,55 @@
+
+![Langflow Environments](/img/langflow-env.png)
+
+* **The Langflow Development environment** must include the Integrated Development Environment (IDE) for full experience of Langflow, optimized for prototyping and testing new flows.
+* **The Langflow Production environment** executes the flow logic in production and productionizes Langflow flows as standalone services.
```
```suggestion
* **The Langflow development environment** must include the Integrated Development Environment (IDE) for the full experience of Langflow, optimized for prototyping and testing new flows.
* **The Langflow production environment** executes the flow logic in production and enables Langflow flows as standalone services.
```

KimberlyFields commented on docs/docs/Deployment/deployment-prod-best-practices.md:
```diff
@@ -0,0 +1,55 @@
+
+* **Security**
+  * **Isolation**: By separating the development and production environments, you can better isolate different phases of the application lifecycle. This isolation minimizes the risk of development-related issues impacting the production environments.
+  * **Access Control**: Different security policies and access controls can be applied to each environment. Developers may require broader access in the IDE for testing and debugging, while the runtime environment can be locked down with stricter security measures.
+  * **Reduced Attack Surface**: The runtime environment is configured to only include essential components, reducing the attack surface and potential vulnerabilities.
+* **Resource Allocation**
+  * **Optimized Resource Usage and cost efficiency**: By separating the two environments, you can allocate resources more effectively. Each flow can be deployed independently, providing fine-grained resource control.
```
```suggestion
  * **Access control**: Different security policies and access controls can be applied to each environment. Developers may require broader access in the IDE for testing and debugging, while the runtime environment can be locked down with stricter security measures.
  * **Reduced attack surface**: The runtime environment is configured to include only essential components, reducing the attack surface and potential vulnerabilities.
* **Resource allocation**
  * **Optimized resource usage and cost efficiency**: By separating the two environments, you can allocate resources more effectively. Each flow can be deployed independently, providing fine-grained resource control.
```

KimberlyFields commented on docs/docs/Develop/develop-application.md:
```diff
@@ -160,4 +160,4 @@ The test application returns a large amount of text, so the example command used
 
-To deploy your application to Kubernetes, see [Kubernetes](/deployment-kubernetes).
\ No newline at end of file
+To deploy your application to Kubernetes, see [Kubernetes](/deployment-kubernetes-prod).
```
```suggestion
To deploy your application to Kubernetes, see [Deploy the Langflow production environment to Kubernetes](/deployment-kubernetes-prod).
```

mendonk reviewed (approved):
LGTM

eric-schneider commented on docs/docs/Deployment/deployment-prod-best-practices.md:
```diff
@@ -0,0 +1,55 @@
+---
+title: Langflow architecture and best practices
```
This page appears to be about Langflow's architecture and best practices _specifically on Kubernetes_, so I'm concerned that this page title is too generic. People doing general searches for "langflow best practices" and "langflow best practices" might be confused when they click on this page title in Google and only find Kubernetes content.

```suggestion
title: Langflow architecture and best practices on Kubernetes
```

KimberlyFields commented on docs/docs/Deployment/deployment-kubernetes-dev.md:
````diff
@@ -0,0 +1,160 @@
+	helm install langflow-ide langflow/langflow-ide -n langflow --create-namespace
+	```
+
+3. Check the status of the pods
````
```suggestion
3. Check the status of the pods.
```

KimberlyFields commented on docs/docs/Deployment/deployment-kubernetes-dev.md:
````diff
@@ -0,0 +1,160 @@
+```
+
+:::note
+If your flow relies on shared state, such as built-in chat memory, you need to set up a shared database when scaling horizontally.
````
```suggestion
If your flow relies on a shared state, such as built-in chat memory, you need to set up a shared database when scaling horizontally.
```

KimberlyFields commented on docs/sidebars.js:
```diff
@@ -112,9 +112,25 @@ module.exports = {
+            {
+              type: "doc",
+              id: "Deployment/deployment-prod-best-practices",
+              label: "Langflow Architecture and Best Practices"
```
```suggestion
              label: "Langflow architecture and best practices"
```

KimberlyFields commented on docs/sidebars.js:
```diff
@@ -112,9 +112,25 @@ module.exports = {
+            {
+              type: "doc",
+              id: "Deployment/deployment-kubernetes-dev",
+              label: "Kubernetes - Development"
```
```suggestion
              label: "Deploy in development"
```

KimberlyFields commented on docs/sidebars.js:
```diff
@@ -112,9 +112,25 @@ module.exports = {
+            {
+              type: "doc",
+              id: "Deployment/deployment-kubernetes-prod",
+              label: "Kubernetes - Production"
```
```suggestion
              label: "Deploy in production"
```

KimberlyFields reviewed (approved):
Left a few suggestions, but approving.
